fix for #19066 Print warnings when deprecated options are configured in config file

mansoor17syed opened the pull request:
fix for #19066 Print warnings when deprecated options are configured in config file

Issue Details
Add warnings when deprecated options are configured
Problem: Currently, etcd doesn't warn users when they configure deprecated options
Goal: Improve user experience by warning about deprecated configurations before they're removed
Solution Overview

Implementation Location: server/config/config.go

serathius commented:
/ok-to-test

ahrtr commented:
This PR doesn't fix #19066 . Please follow https://github.com/etcd-io/etcd/pull/18998#discussion_r1885830028

Thanks anyway

mansoor17syed commented:
Hi @ahrtr 

Thank you for the review and feedback!

I initially assumed that the comment in the earlier PR referred specifically to their changes. Could you please confirm if the fix for issue #19066 is intended to address the comment you provided here : https://github.com/etcd-io/etcd/pull/18998#discussion_r1885830028

Appreciate your clarification!

ahrtr commented:
> Could you please confirm if the fix for issue #19066 is intended to address the comment you provided here : [#18998 (comment)](https://github.com/etcd-io/etcd/pull/18998#discussion_r1885830028)

YES

ahrtr commented:
I just realised that we can iterate `FlagsExplicitlySet` directly in [parse()](https://github.com/etcd-io/etcd/blob/75f2ae1b6280cd16dde122ae55e815e3fce5a4e8/server/etcdmain/config.go#L128).  Both `configFromFile` and `configFromCmdLine` populate the data structure with all flags,

https://github.com/etcd-io/etcd/blob/75f2ae1b6280cd16dde122ae55e815e3fce5a4e8/server/embed/config.go#L486


sorry for the back and forth.

mansoor17syed commented:
@ahrtr 

Can you Please have a look once now if I have done the right changes 
I see one tests failing , but I guess the failure is not because of my changes 

Still open for feedback and suggestions
do let me know

ahrtr commented:
@mansoor17syed The change looks good.  Please squash the commits

mansoor17syed commented:
@ahrtr 

can chk now I have added the UT but some seems failing I doubt its because of my changes.
Still open for feedback and suggestions
do let me know

ahrtr commented on server/etcdmain/config_test.go:
```diff
@@ -653,3 +654,101 @@ func validateClusteringFlags(t *testing.T, cfg *config) {
 }
+
+func TestConfigFileDeprecatedOptions(t *testing.T) {
+	type configStruct struct {
+		SnapshotCount                           uint64        `json:"snapshot-count,omitempty"`
+		MaxSnapFiles                            uint          `json:"max-snapshots,omitempty"`
+		V2Deprecation                           string        `json:"v2-deprecation,omitempty"`
+		ExperimentalCompactHashCheckEnabled     bool          `json:"experimental-compact-hash-check-enabled,omitempty"`
+		ExperimentalCompactHashCheckTime        time.Duration `json:"experimental-compact-hash-check-time,omitempty"`
+		ExperimentalWarningUnaryRequestDuration time.Duration `json:"experimental-warning-unary-request-duration,omitempty"`
+	}
```
I think that you can reuse [embed.Config](https://github.com/etcd-io/etcd/blob/9db8dcb44b9ad7d0cf94438169cef7a778e2eb15/server/embed/config.go#L162) directly instead of defining another struct.

ahrtr commented on server/etcdmain/config_test.go:
```diff
@@ -653,3 +654,101 @@ func validateClusteringFlags(t *testing.T, cfg *config) {
+			// Compare sets of flags
+			if !reflect.DeepEqual(foundFlags, tc.expectedFlags) {
+				t.Errorf("deprecated flags mismatch:\ngot: %v\nwant: %v",
+					mapToSortedSlice(foundFlags),
+					mapToSortedSlice(tc.expectedFlags))
```
```suggestion
					slices.Sorted(maps.Keys(foundFlags)),
					slices.Sorted(maps.Keys(tc.expectedFlags))
```

ahrtr commented on server/etcdmain/config_test.go:
```diff
@@ -653,3 +654,101 @@ func validateClusteringFlags(t *testing.T, cfg *config) {
+	}
+}
+
+// Helper function to convert map keys to sorted slice for better error messages
+func mapToSortedSlice(m map[string]struct{}) []string {
+	result := make([]string, 0, len(m))
+	for k := range m {
+		result = append(result, k)
+	}
+	sort.Strings(result)
+	return result
+}
```
```suggestion
```

ahrtr commented:
Please squash the commits.

mansoor17syed commented:
@ahrtr @serathius 
can you Please have a look  into it now, addressed all the review comments and squashed the commits.

mansoor17syed commented:
I don’t seem to have the permissions to merge this. @ahrtr  Could you kindly take care of it? 
Thanks so much for all your support and valuable feedback! 😊

ahrtr commented:
Usually a PR needs two approval to be merged. Let's wait one of other maintainers or reviewers to take a look. thx

serathius commented on server/etcdmain/config_test.go:
```diff
@@ -653,3 +655,89 @@ func validateClusteringFlags(t *testing.T, cfg *config) {
+			}
+
+			// Compare sets of flags
+			if !reflect.DeepEqual(foundFlags, tc.expectedFlags) {
```
Why not use assert.Equal?

serathius commented on server/etcdmain/config_test.go:
```diff
@@ -653,3 +655,89 @@ func validateClusteringFlags(t *testing.T, cfg *config) {
+
+			// Special check for experimental-warning-unary-request-duration
+			if tc.configFileYAML.ExperimentalWarningUnaryRequestDuration != 0 {
+				t.Log("Note: experimental-warning-unary-request-duration deprecation is handled separately")
```
Why log here? if we want to leave context a comment would be enough.

mansoor17syed replied:
Ok will chk this

mansoor17syed commented:
@serathius 

can you have a look now ?
Do let me know for any rework

serathius commented:
```
etcdmain/config_test.go:30: File is not `goimports`-ed with -local go.etcd.io (goimports)
	"github.com/stretchr/testify/assert"
etcdmain/config_test.go:734:4: SA9003: empty branch (staticcheck)
			if tc.configFileYAML.ExperimentalWarningUnaryRequestDuration != 0 {
			^
etcdmain/config_test.go:730:4: formatter: use assert.Equalf (testifylint)
			assert.Equal(t, tc.expectedFlags, foundFlags, "deprecated flags mismatch")
```

mansoor17syed commented:
@serathius 

I have addressed the review comments and fixed the issues. However, it seems some tests are still failing, although they don’t appear to be related to my changes. I'm open to any feedback and willing to work on it further if needed. Let me know your thoughts!

mansoor17syed commented:
I don’t seem to have the permissions to merge this. @serathius  Could you kindly take care of it?
Thanks so much for all your support and valuable feedback! 😊
